### PR TITLE
Boilerplate packaging stuff

### LINKED ISSUE #11: Packaging
@miguelcleon I am unfamiliar with packages that use Djando so I don't really know where to start. Usually the boilerplate is to create a `setup.py` that installs the module, but I believe there are more pieces here that needs to be taken care of. Is that assessment correct?

xref: https://github.com/ODM2/conda-recipes-ODM2/issues/27

### PR DESCRIPTION
**Pull request by ocefpaf**
This is not ready to be merge. It is just to start the discussing the `odm2djangoadmin` packaging.

(Closes #11)

### CODE REVIEW

File: LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Copyright (c) 2015
```
Comment (ocefpaf) [inline LICENSE.txt:1]: @miguelcleon This is the `MIT` License. I could not find one, but if you do not want to use MIT let me know and I will change this file.

File: MANIFEST.in [hunk 2]
```diff
@@ -0,0 +1,4 @@
+include README.rst
+include *.txt
+recursive-include odm2djangoadmin *.py
+recursive-include odm2djangoadmin *.html
```
Comment (ocefpaf) [inline MANIFEST.in:4]: I never package Django apps before. I need to check if these need to be package as data files and deployed somewhere at run time.

File: README.md [hunk 3]
```diff
@@ -1,26 +0,0 @@
-
```
Comment (ocefpaf) [inline README.md:1]: I removed all the `pyc` (above) as they are by-products from a previous run and should not be in the repository.

I also converted the `README.md` (Markdown) to `README.rst` (ReST). GitHub can render both, but PyPI can only render ReST.

File: odm2djangoadmin/__init__.py [hunk 4]
```diff
@@ -0,0 +1 @@
+__version__ = '0.1.0'
```
Comment (ocefpaf) [inline odm2djangoadmin/__init__.py:1]: I moved all the python files 1-level up and left the packaging metadata in the root directory. That is common practice in most python packages, but I am still reading the proper way to do a `django` app.

---

Edit: also add an initial version number.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      url='https://github.com/miguelcleon/odm2djangoadmin',
+      platforms='any',
+      keywords=['ODM2', 'Django'],
+      install_requires=install_requires,
+      packages=['odm2djangoadmin'],
+      tests_require=['pytest'],
+      cmdclass=dict(test=PyTest),
+      author=['Miguel Leon'],
+      author_email='leonmi@sas.upenn.edu',
+      )
```
Comment (ocefpaf) [inline setup.py:75]: This is a typical boilerplate `setup.py` file. @miguelcleon can you review the metadata?

Note that:
- the PyPI description will be the `README.rst` and the `LICENSE.txt` files;
- the version is extracted from the `__init__.py` file;
- there are no tests but `setup.py` can be run with `python setup.py test` for auto-discovery of pytests once we implemente them;
- `python setup.py sdist` creates the tarball;
- `python setup.py build` build the install tree (good for package tests);
- `python setup.py install` is used to install the package (although nowadays it is recommend to use `pip` even for local installs);
- `python setup.py register` register at PyPI;
- `python setup.py sdist --formats=gztar upload` uploads to PyPI.

**Comment by ocefpaf** [comment]
@miguelcleon I would like to set a call with you so I can ask a few quick questions about how you run the app so I can figure out a strategy to complete this PR.

File: .travis.yml [hunk 6]
```diff
@@ -1,7 +1,28 @@
 language: python
+
+sudo: false
```
Comment (ocefpaf) [inline .travis.yml:3]: Travis-CI will run faster if we set sudo to `false`.

File: .travis.yml [hunk 7]
```diff
@@ -1,7 +1,28 @@
...
 python:
-  - "2.7"
-# command to install dependencies
-install: 
- - "pip install -r requirements.txt"
- #- "pip install git+https://github.com/nwcell/psycopg2-windows.git@win64-py27#egg=psycopg2" #for windows, travis is building on ubuntu
\ No newline at end of file
+    - 2.7
+    - 3.4
+    - 3.5
```
Comment (ocefpaf) [inline .travis.yml:8]: Not sure if the code is compatible with Python 3 yet, but it does not hurt to test.

File: .travis.yml [hunk 8]
```diff
@@ -1,7 +1,28 @@
...
+
+env:
+  - TEST_TARGET=default
+  - TEST_TARGET=coding
```
Comment (ocefpaf) [inline .travis.yml:12]: I split the test into to contexts. Actual tests and coding standards. We can set coding standards to `allow_failures` to reduce the noise and get a green Travis-CI, but I like to enforce it :grimacing:

File: .travis.yml [hunk 9]
```diff
@@ -1,7 +1,28 @@
...
+
+before_install:
+    - travis_retry pip install -r requirements.txt
+
+# Test the tarball creation.
+install:
+    - python setup.py sdist && version=$(python setup.py --version) && pushd dist  && pip install odm2djangoadmin-${version}.tar.gz && popd
```
Comment (ocefpaf) [inline .travis.yml:19]: This is the "same" test as you already have. Just test the tarball creation and install it. There are no real tests yet.

File: .travis.yml [hunk 10]
```diff
@@ -1,7 +1,28 @@
...
+
+script:
+    # Un-comment once we have proper testing.
+    #- if [[ $TEST_TARGET == 'default' ]]; then
+        #python setup.py test -a '--verbose --cov=pyugrid';
+      #fi
```
Comment (ocefpaf) [inline .travis.yml:25]: We should un-comment this block once we have real tests.

**Comment by emiliom** [comment]
Great, @ocefpaf! FYI, we have other people in the ODM2 team experienced with Django (but not conda packaging), if you and @miguelcleon would like additional input.

File: README.md (hunk 3 above)
Comment (miguelcleon) [inline README.md:1]: Agreed.

**Comment by miguelcleon** [comment]
Hi, yeah we can have a call let me know when would be good.

**Comment by ocefpaf** [comment]
>  if you and @miguelcleon would like additional input.

It is always nice to get extra input.

@miguelcleon I need to add a few extra commits to ship the `HTML` parts as package data and then we can have a call on how you currently run `odm2djangoadmin`.

**Comment by miguelcleon** [comment]
ok

**Comment by emiliom** [comment]
> > if you and @miguelcleon would like additional input.
> 
> It is always nice to get extra input.

Looks like you two have a good handle on this. If you do want third-party Django eyeballs for something specific, let me know and I'll try to recruit them.

**Comment by miguelcleon** [comment]
How is this coming along? Need anything from me?

File: .travis.yml (hunk 7 above)
Comment (miguelcleon) [inline .travis.yml:8]: Probably won't work with Python 3 right now, but yeah doesn't hurt to test.

File: odm2djangoadmin/__init__.py (hunk 4 above)
Comment (miguelcleon) [inline odm2djangoadmin/__init__.py:1]: Yeah, I don't think this django app will work that way. Have you gotten it to run like this?

File: setup.py (hunk 5 above)
Comment (miguelcleon) [inline setup.py:75]: metadata seems fine.

File: odm2djangoadmin/__init__.py (hunk 4 above)
Comment (ocefpaf) [inline odm2djangoadmin/__init__.py:1]: Sorry @miguelcleon I haven't got to it yet. I am a little but swamped with some conda-forge packages. I will get back to this before the end of the week.

**Comment by ocefpaf** [comment]
> How is this coming along? Need anything from me?

Nope. I am getting close to cut a release that works as expected. I just need to smooth the static data install.

**Comment by miguelcleon** [comment]
Ok, great, let me know when you think it's ready to go.

**Comment by miguelcleon** [comment]
hi @ocefpaf how's it coming? Sorry about the confusion about django-admin.py, all of the Django CLI commands should be run through manage.py

**Comment by miguelcleon** [comment]
hi @ocefpaf how's it coming?

**Comment by ocefpaf** [comment]
Hi @miguelcleon this is on halt. I Wrote @emiliom today to schedule a call so I can touch base with him and sort some details out.

**Comment by ocefpaf** [comment]
I will refresh this PR using latest master.

**Comment by miguelcleon** [comment]
ok sounds good.

**Pull request closed without merging**